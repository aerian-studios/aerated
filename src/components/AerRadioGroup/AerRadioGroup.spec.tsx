import React from "react";
import { describe, it, expect, afterEach, vi } from "vitest";
import { render, cleanup, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { AerRadioGroup } from "./index";

describe("AerRadioGroup", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render correctly", async () => {
    const { container } = render(
      <AerRadioGroup.Root value="docs" groupLabel={"Subject under test"}>
        <AerRadioGroup.Item label="Testing" value="testing" checked />
        <AerRadioGroup.Item label="Writing docs" value="docs" />
        <AerRadioGroup.Item label="Code reviews" value="pr" />
        <AerRadioGroup.Item
          label="Writing new features"
          disabled={true}
          value="new"
        />
      </AerRadioGroup.Root>
    );

    // using implicit assertion of existence inherent in selection
    screen.getByRole("group", { name: "Subject under test" });

    screen.getByRole("radio", { name: "Testing" });
    screen.getByRole("radio", { name: "Writing docs" });
    screen.getByRole("radio", { name: "Code reviews" });
    screen.getByRole("radio", { name: "Writing new features" });

    expect(container).toMatchSnapshot();
  });

  it("should control the selected radio based on the Root `value` prop", async () => {
    render(
      <AerRadioGroup.Root value="docs" groupLabel={"Subject under test"}>
        <AerRadioGroup.Item label="Testing" value="testing" checked />
        <AerRadioGroup.Item label="Writing docs" value="docs" />
        <AerRadioGroup.Item label="Code reviews" value="pr" />
        <AerRadioGroup.Item
          label="Writing new features"
          disabled={true}
          value="new"
        />
      </AerRadioGroup.Root>
    );

    expect(
      (screen.getByRole("radio", { name: "Testing" }) as HTMLInputElement)
        .checked
    ).toBeFalsy();
    expect(
      (screen.getByRole("radio", { name: "Code reviews" }) as HTMLInputElement)
        .checked
    ).toBeFalsy();
    expect(
      (screen.getByRole("radio", { name: "Writing docs" }) as HTMLInputElement)
        .checked
    ).toBeTruthy();

    await userEvent.click(screen.getByRole("radio", { name: "Testing" }));

    expect(
      (screen.getByRole("radio", { name: "Testing" }) as HTMLInputElement)
        .checked
    ).toBeTruthy();
  });

  it.only("should call the `onChange` functions provided with the correct values", async () => {
    const mockGroupChange = vi.fn();
    const mockRadioChange = vi.fn();
    render(
      <AerRadioGroup.Root
        value="new"
        groupLabel={"Subject under test"}
        onChange={mockGroupChange}
      >
        <AerRadioGroup.Item
          label="Testing"
          value="testing"
          onChange={mockRadioChange}
        />
        <AerRadioGroup.Item label="Writing docs" value="docs" />
        <AerRadioGroup.Item
          label="Writing new features"
          disabled={true}
          value="new"
          onChange={mockRadioChange}
        />
      </AerRadioGroup.Root>
    );

    expect(mockGroupChange).not.toHaveBeenCalled();
    expect(mockRadioChange).not.toHaveBeenCalled();

    const testingRadio: HTMLInputElement = screen.getByRole("radio", {
      name: "Testing",
    });
    await userEvent.click(testingRadio);

    expect(mockGroupChange).toHaveBeenCalledTimes(1);
    expect(mockGroupChange).toHaveBeenCalledWith("testing");
    expect(mockRadioChange).toHaveBeenCalledTimes(1);
    expect(mockRadioChange).toHaveBeenCalledWith(
      expect.objectContaining({ type: "change", target: testingRadio })
    );
    expect(testingRadio.checked).toBeTruthy();

    // this is disabled
    await userEvent.click(
      screen.getByRole("radio", { name: "Writing new features" })
    );

    expect(mockGroupChange).not.toHaveBeenCalledWith("new");
    expect(mockGroupChange).toHaveBeenCalledTimes(1);
    expect(mockRadioChange).toHaveBeenCalledTimes(1);

    expect(
      (
        screen.getByRole("radio", {
          name: "Writing new features",
        }) as HTMLInputElement
      ).checked
    ).toBeFalsy();

    // not disabled
    const docsRadio: HTMLInputElement = screen.getByRole("radio", {
      name: "Writing docs",
    });
    await userEvent.click(docsRadio);

    expect(mockGroupChange).toHaveBeenCalledTimes(2);
    expect(mockGroupChange).toHaveBeenCalledWith("docs");
    expect(mockRadioChange).toHaveBeenCalledTimes(1);
    expect(mockRadioChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: docsRadio,
      })
    );
    expect(docsRadio.checked).toBeTruthy();
  });
});
