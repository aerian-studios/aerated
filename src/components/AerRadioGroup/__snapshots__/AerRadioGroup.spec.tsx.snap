// Vitest Snapshot v1

exports[`AerRadioGroup > should render correctly 1`] = `
<div>
  <fieldset
    class="_component_61bd41"
    role="radiogroup"
  >
    <legend
      class="_title_61bd41"
    >
      Subject under test
    </legend>
    <label
      class="_radioInput_61bd41"
    >
      <span
        class="_indicator_61bd41"
      >
        <input
          class="_input_61bd41"
          name="Subject-under-test"
          role="radio"
          type="radio"
        />
        <span
          aria-hidden="true"
          class="_imgWrapper_61bd41 _radioBackground_61bd41"
        />
      </span>
      <span
        class="_label_61bd41"
      >
        Testing
      </span>
    </label>
    <label
      class="_radioInput_61bd41"
    >
      <span
        class="_indicator_61bd41"
      >
        <input
          checked=""
          class="_input_61bd41"
          name="Subject-under-test"
          role="radio"
          type="radio"
        />
        <span
          aria-hidden="true"
          class="_imgWrapper_61bd41 _radioBackground_61bd41"
        >
          <svg
            class="_checkedIcon_61bd41"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.875 7.5C9.875 8.81168 8.81168 9.875 7.5 9.875C6.18832 9.875 5.125 8.81168 5.125 7.5C5.125 6.18832 6.18832 5.125 7.5 5.125C8.81168 5.125 9.875 6.18832 9.875 7.5Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
      <span
        class="_label_61bd41"
      >
        Writing docs
      </span>
    </label>
    <label
      class="_radioInput_61bd41"
    >
      <span
        class="_indicator_61bd41"
      >
        <input
          class="_input_61bd41"
          name="Subject-under-test"
          role="radio"
          type="radio"
        />
        <span
          aria-hidden="true"
          class="_imgWrapper_61bd41 _radioBackground_61bd41"
        />
      </span>
      <span
        class="_label_61bd41"
      >
        Code reviews
      </span>
    </label>
    <label
      class="_radioInput_61bd41"
      data-disabled="true"
    >
      <span
        class="_indicator_61bd41"
      >
        <input
          class="_input_61bd41"
          disabled=""
          name="Subject-under-test"
          role="radio"
          type="radio"
        />
        <span
          aria-hidden="true"
          class="_imgWrapper_61bd41 _radioBackground_61bd41"
        />
      </span>
      <span
        class="_label_61bd41"
      >
        Writing new features
      </span>
    </label>
  </fieldset>
</div>
`;
